index.js

```diff
@@ -210,7 +210,7 @@ function isRequestOriginAllowed (reqOrigin, allowedOrigin) {
   } else if (typeof allowedOrigin === 'string') {
     return reqOrigin === allowedOrigin
   } else if (allowedOrigin instanceof RegExp) {
-    return allowedOrigin.test(reqOrigin)
+    return Boolean(reqOrigin.match(allowedOrigin))
```

Comment:
can't we just do?
```suggestion
    allowedOrigin.lastIndex=0
    return allowedOrigin.test(reqOrigin)
```

Reply:
@zekth I don't know, I'm not sure about the exact implementation of Node.js async/thread behaviour. E.g. if two requests are triggered simultaneously, can we ensure that the following does not occur?:
1. First request is running .test function
2. Second request sets 'lastIndex' to 0
3. First requests finds match and updates lastIndex again
4. Second requests then uses the updated lastIndex

If we are sure the above cannot happen, and these lines are always executed synchronously without interruption, we can also use your solution since it's less invasive.

Please let me know what you think, then I'll update the commit...

Reply:
This cannot happen as we are not working with multithreading but asynchronization. As those operations happen in the same scope there would not be any collision regarding the change of the regex.

On the other hand this shows a problem that might appear, which is passing a reference of the regex in the plugin like:
```js
const reg = /(example|other)\.com/gi
fastify.register(cors, { origin: reg })
```
In this case, if `reg` is mutated anywhere else in the app we'll have a collision. So in this case the only way to avoid this is to do something like
```js
if (allowedOrigin instanceof RegExp) {
    const testReg = RegExp(allowedOrigin.source, allowedOrigin.flags)
    return testReg.test(reqOrigin)
}
```

Reply:
You are correct. I just tested the performance implications of three variants:
```javascript
const sourceRegex = new RegExp('example.com', 'gi');
const testString = 'https://www.example.com/';

function timeRegular() { // Takes about 33ms
  start = performance.now();
  for (let i=0; i < 1000000; i++) {
    sourceRegex.lastIndex = 0;
    sourceRegex.test(testString);
  }
  end = performance.now()
  return end - start;
}

function timeCopy() { // Takes ~ 160 ms
  start = performance.now();
  for (let i=0; i < 1000000; i++) {
    const testRegex = new RegExp(sourceRegex.source, sourceRegex.flags);
    testRegex.test(testString);
  }
  end = performance.now()
  return end - start;
}

function timeMatch() { // Takes ~100ms
  start = performance.now();
  for (let i=0; i < 1000000; i++) {
    testString.match(sourceRegex);
  }
  end = performance.now()
  return end - start;
}
```

The copying seems to add some overhead to the process (factor 5 roughly), but for a million checks it's only about 130ms extra.

I also checked [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp), and it seems like 'lastIndex' is the only editable attribute for a RegExp object.

If you think this additional overhead compares positively to being sure that 'lastIndex' is zero at the start, please let me know, I'll implement your last suggestion.

Personally I would go with your first suggestion (Which is also the fastest from the above tests), since I wouldn't even want to promote being able to count on the 'lastIndex' staying zero anywhere in the plugin, but that may be just my personal preference...

Reply:
@climba03003 I would like to have your opinion on this.

Reply:
I will go for editing the `lastIndex`. If you think it may alter by someone in the middle of run. You can clone it once before passing down to the check. It should be in https://github.com/fastify/fastify-cors/pull/152/files#R101